database/multistmt/parse.go

```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 	// completed statements, contents of accum will be dumped in here
 	stmts := make([][]byte, 0, 1000)
 
+	tmp := make([]byte, 0, 10)
+	a := 0
 	for err == nil {
+		// this is non-performant but very clear that the buf variable is wiped out
+		//completely between loop iterations
+		buf = make([]byte, ParseBufSize)
 		n, err := reader.Read(buf)
+		trace("tmp(2): '%s', buf: %s, discard: %v\n", tmp, buf, discard)
+		// tmp is the carry-over buffer,
+		//if the previous loop iteration had two few characters to make comparisions,
+		//tmp will have the characters at the point the loop iteration was abandoned(
+		//break'd out of)
+		if len(tmp) > 0 {
+			trace("copying '%s' to buf\n", tmp)
+			buf = append(tmp, buf[:n]...)
+			trace("buf: %s\n", buf)
+			// n needs to include the length of tmp since it was copied into n,
+			// but n originally only held the number of chars read from the
+			// reader.Read(buf) call.
+			n = n + len(tmp)
+			tmp = tmp[:0]
+
+		}
 		if n > 0 {
+			// buf needs capacity(it is initialized with capapcity and length the same)
+			// so we can only loop to the bytes read, not the capacity nor length
+			// there may also be bytes copied over from the previous loop interation
+			// that are now in buf also.
 			for i := range buf[:n] {
+				// 2 here is the number of look ahead characters that we use.
+				// This tmp buffer is used to copy over bytes from the current loop
+				// iteration if there are not enough characters to lookahead and find a match
+				if i+1 >= len(buf) {
+					tmp = make([]byte, n-i)
```

Comment:
related to my question above, isn't `tmp` always going to be size of 1? (Since we only need to read ahead 1 additional character).

If so, could we do this to make it simpler to read?

```suggestion
					tmp = make([]byte, 1)
```

Reply:
Only if we never do comparisons looking for two additional characters, I'd like to keep the flexibility